src/library/dao/trackdao.cpp

```diff
@@ -1622,14 +1622,16 @@ void TrackDAO::markTrackLocationsAsDeleted(const QString& directory) {
     }
 }
 
-// Look for moved files. Look for files that have been marked as "deleted on disk"
-// and see if another "file" with the same name and filesize exists in the track_locations
-// table. That means the file has moved instead of being deleted outright, and so
-// we can salvage your existing metadata that you have in your DB (like cue points, etc.).
-void TrackDAO::detectMovedFiles(QSet<int>* pTracksMovedSetOld, QSet<int>* pTracksMovedSetNew) {
-    //This function should not start a transaction on it's own!
-    //When it's called from libraryscanner.cpp, there already is a transaction
-    //started!
+// Look for moved files. Look for files that have been marked as
+// "deleted on disk" and see if another "file" with the same name and
+// files size exists in the track_locations table. That means the file has
+// moved instead of being deleted outright, and so we can salvage your
+// existing metadata that you have in your DB (like cue points, etc.).
+bool TrackDAO::detectMovedFiles(QSet<int>* pTracksMovedSetOld,
+        QSet<int>* pTracksMovedSetNew, volatile const bool* pCancel) {
```

Comment:
> But then we should rather improve this function.

Yes of cause. Since I have not a good and easy receipted yet we should postpone it after 1.12. 

During this PR I have tried careful not to change what happens by a cancel. It should be exactly the same as in current 1.12, except that cancelling during the clean up phase it excepted an the result should be the same as if you have cancelled the scan just before. 

> Currently canceling means that only minimal progress is lost.

These are he lines questions (untouched by this PR): 
```
    // Finish adding the tracks -- rollback the transaction if the scan did not
    // finish cleanly and the user did not cancel the transaction.
    m_trackDao.addTracksFinish(!m_scannerGlobal->shouldCancel() &&
                               !bScanFinishedCleanly); 
```
tracks from a manual cancelled scan are still added to the library.  

The cleanup tasks are already skipped in 1.12 by the following expression, when you cancel the can during the folder scan phase. 
```
if (bScanFinishedCleanly) {
```

You see that the result should be exactly the same as if we cancel just after the final folder is scanned. Can you confirm this? 

> Your solution might fix the long shutdown time of mixxx in canceled scans but I don't think some one would rescan his entire library during a performance. The better solution in the long run is to improve the performance of this.

Remember that Mixxx automatically starts a rescan in some cases. 
And Yes, it is stupid to do a scan during a performance and once someone notice it he should be able to mediately cancel it. I was myself that stupid, because I could not find a hardly needed track.

An improved scanner performance is only an addition to a cancel function that works almost mediately. 
Even if we manage to reduce the clean up time to 10 % (which will be unlikely possible) we are still talking about 10 long second on a recent device which IMHO require a cancel. 

> It should also be possible to write a unit-test for the clean-up functions. They only depend on internal library state.

Which test cases do you have in mind? What should be verified?